src/software/ai/hl/stp/tactic/goalie/goalie_fsm.h

```diff
@@ -0,0 +1,448 @@
+
+        // if the goalie could not be restrained in the defense area,
+        // then the ball must be either on a really sharp angle to the net where
+        // its impossible to get a shot, or the ball is behind the net, in which
+        // case we snap to either post
+        if (!clamped_goalie_pos)
+        {
+            if (ball.position().y() > 0)
+            {
+                goalie_pos =
+                    field.friendlyGoalpostPos() + Vector(0, -ROBOT_MAX_RADIUS_METERS);
+            }
+            else
+            {
+                goalie_pos =
+                    field.friendlyGoalpostNeg() + Vector(0, ROBOT_MAX_RADIUS_METERS);
+            }
+        }
+        else
+        {
+            goalie_pos = *clamped_goalie_pos;
+        }
+        return goalie_pos;
+    }
+
+    /**
+     * Gets intersections between the ball velocity ray and the full goal segment
+     *
+     * @param ball the ball to find interceptions with the full goal segment
+     * @param field the field to find interceptions on
+     *
+     * @return the intersections between the ball velocity ray and the full goal segment
+     */
+    static std::vector<Point> getIntersectionsBetweenBallVelocityAndFullGoalSegment(
+        const Ball &ball, const Field &field)
+    {
+        // compute intersection points from ball position and velocity
+        Ray ball_ray = Ray(ball.position(), ball.velocity());
+
+        // Create a segment along the goal line, slightly shortened to account for the
+        // robot radius so as we move along the segment we don't try to run into the goal
+        // posts. This will be used in case 3 as a fallback when we don't have an
+        // intersection with the crease lines
+        Segment full_goal_segment =
+            Segment(field.friendlyGoalpostNeg() + Vector(0, -ROBOT_MAX_RADIUS_METERS),
+                    field.friendlyGoalpostPos() + Vector(0, ROBOT_MAX_RADIUS_METERS));
+
+        return intersection(ball_ray, full_goal_segment);
+    }
+
+    /**
+     * Gets the "don't chip" rectangle inside the friendly defense area
+     *
+     * @param field the field to find the "don't chip" rectangle on
+     *
+     * @return the "don't chip" rectangle
+     */
+    static Rectangle getDontChipRectangle(const Field &field)
+    {
+        return Rectangle(
+            field.friendlyGoalpostNeg(),
+            field.friendlyGoalpostPos() + Vector(2 * ROBOT_MAX_RADIUS_METERS, 0));
+    }
+
+    /**
+     * Gets the point on the field from which the goalie should clear the ball
+     * @param field the field on which the goalie will clear the ball
+     * @param origin_opt the optional point to clear the ball from
+     *
+     * @return the position that the goalie should clear the ball from
+     */
+    static Point getClearOrigin(const Field &field, std::optional<Point> origin_opt)
+    {
+        Point clear_origin = field.friendlyGoalCenter() + Vector(0.5, 0);
```

Comment:
Yeah I can definitely change this, this was just to make sure the goalie was outside of the no-chip rectangle - also I need to consider how far out I should set this value since the goalie wouldn't be able to dribble more than 1 m (I think this could potentially be an issue, depending where I choose the chip origin?)

Reply:
Perhaps the clearorigin can depend on the no-chip rectangle then?

Reply:
Hmm ok, so maybe something like the goalie dribbles in the positive-X direction until it is outside of the no-chip rectangle?

Reply:
yeah the clear origin could be based on `no_chip_rectangle.xMax()`. Also, I like the sound of no-chip rectangle over dont chip rectangle

Reply:
updated